hazelcast/src/main/java/com/hazelcast/jet/impl/metrics/JobMetricsPublisher.java

```diff
@@ -29,56 +30,59 @@
 import java.util.Map;
 import java.util.Objects;
 import java.util.Set;
+import java.util.function.BiConsumer;
 import java.util.function.UnaryOperator;
 
 /**
- * A publisher which updates the latest metric values in {@link
- * JobExecutionService}.
+ * A publisher which updates the latest metric values in
+ * {@link JobCoordinationService} and {@link JobExecutionService}.
  */
 public class JobMetricsPublisher implements MetricsPublisher {
-
+    private final JobCoordinationService jobCoordinationService;
     private final JobExecutionService jobExecutionService;
     private final UnaryOperator<MetricDescriptor> namePrefixFn;
+    private final Map<Long, MetricsCompressor> jobIdToCompressor = new HashMap<>();
     private final Map<Long, MetricsCompressor> executionIdToCompressor = new HashMap<>();
 
     public JobMetricsPublisher(
+            @Nonnull JobCoordinationService jobCoordinationService,
             @Nonnull JobExecutionService jobExecutionService,
             @Nonnull Member member
     ) {
+        Objects.requireNonNull(jobCoordinationService, "JobCoordinationService");
         Objects.requireNonNull(jobExecutionService, "jobExecutionService");
         Objects.requireNonNull(member, "member");
 
+        this.jobCoordinationService = jobCoordinationService;
         this.jobExecutionService = jobExecutionService;
         this.namePrefixFn = JobMetricsUtil.addMemberPrefixFn(member);
     }
 
     @Override
     public void publishLong(MetricDescriptor descriptor, long value) {
-        MetricsCompressor metricsCompressor = getCompressor(descriptor);
-        if (metricsCompressor != null) {
-            metricsCompressor.addLong(namePrefixFn.apply(descriptor), value);
-        }
+        publish(descriptor, (compressor, desc) -> compressor.addLong(desc, value));
     }
 
     @Override
     public void publishDouble(MetricDescriptor descriptor, double value) {
-        MetricsCompressor metricsCompressor = getCompressor(descriptor);
-        if (metricsCompressor != null) {
-            metricsCompressor.addDouble(namePrefixFn.apply(descriptor), value);
-        }
+        publish(descriptor, (compressor, desc) -> compressor.addDouble(desc, value));
     }
 
     @Override
     public void whenComplete() {
-        Set<Map.Entry<Long, MetricsCompressor>> compressorEntries = executionIdToCompressor.entrySet();
+        update(jobIdToCompressor, jobCoordinationService::updateMetrics);
+        update(executionIdToCompressor, jobExecutionService::updateMetrics);
+    }
+
+    private void update(Map<Long, MetricsCompressor> compressorMap, BiConsumer<Long, RawJobMetrics> updateFn) {
```

Comment:
this needs further scrutiny if it is thread safe (jobs can be started concurrently to metrics collection) and if compressors do not leak to not reintroduce #24225